### DISCUSSION TITLE
CMS features & architecture ?

### DISCUSSION THREAD

**Discussion by fbaube** [pocketbase/pocketbase#50] [Q&A]
A good application area for pocketbase might be to provide infrastructural support for a CMS  (Content Management System).
If I understand the documentation correctly, pocketbase shares an important feature with MS SharePoint: for an existing collection, additional property fields (record attributes) can be added & dropped, and can be strongly typed (email, url, etc.). This flexibility in SharePoint is quite valuable for a CMS. Also pocketbase supports notification triggers on collections and on individual records, which can also be quite valuable in a CMS.
I wonder though, what would a CMS "add-on" for pocketbase look like ? It might include additional DB tables, and useful enumerations (such as "workflow state" and other facets), and a set of DB event triggers, and maybe some UI templates. Perhaps a WYSIWYG editor like prosemirror or tiptap, that works for HTML and Markdown and might also be configured with a custom schema. Would there be an easy way to bundle all these elements as a single pocketbase "add-on" that could be installed smoothly ? 
I understand that this is thinking aloud, and I apologize if this topic is out of scope or very ill-defined.

**Comment by ganigeorgiev** [comment]
While some CMS functionalities could be added to the existing admin UI, it really wasn't designed for such tasks as there aren't even admin access controls (all admins act as "super admin" for the application; you can apply access controls only for the users). I've created the admin UI primarily for simple application configurations and quick data explorations (similar to the Firebase console dashboard).

My recommendation for more complex CMS features would be to use [PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and embed your own client-side application that will act as the CMS admin.

**Reply by fbaube** [reply]
Oh sure. That's basically what I have in mind. So I am wondering how to do this as "seamlessly" as possible, so that the CMS functionality works _with_ pocketbase rather than _against_ it. Some kind of "add-on" that sits atop pocketbase-the-framework.

**Reply by ganigeorgiev** [reply]
There is no need to be as "add-on", they can be used together and still have a single self-contained and portable binary (you could think of PocketBase as Go library/toolkit).

The easiest way in my opinion would be to have a static SPA application that can be embedded in your go program (similar to how we currently embed the admin UI) and use the [event hooks](https://pocketbase.io/docs/event-hooks/) to modify the default PocketBase behavior.
I'm planning to do exactly this in my other open source project, which is actually how the initial idea for PocketBase came up first (you can read more in [Presentator #183](https://github.com/presentator/presentator/issues/183)).

**Reply by marcusfelix** [reply]
@ganigeorgiev how exactly we can override the default admin UI located at `/_`? I tried with a custom API route and a `/_` folder inside "pb_public" without success.

**Reply by ganigeorgiev** [reply]
@marcusfelix Do you mean you want to hide/replace the Admin UI at `/_/`?
You could "override" or "replace" any routes with a middleware but you'll have to use [PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and build your executable. For routing we are using [**echo v5**](https://github.com/labstack/echo/tree/v5_alpha):

```go
package main

import (
	"log"
	"net/http"
	"strings"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
)

func adminRedirect(app core.App) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			if strings.HasPrefix(c.Request().URL.Path, "/_/") {
				// admin page -> redirect to your custom url
				return c.Redirect(http.StatusTemporaryRedirect, "https://pocketbase.io")
			}

			return next(c)
		}
	}
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.Pre(adminRedirect(app))

		// serves static files from the provided public dir (if exists)
		subFs := echo.MustSubFS(e.Router.Filesystem, "pb_public")
		e.Router.GET("/*", apis.StaticDirectoryHandler(subFs, false))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

But usually you shouldn't override the default Admin UI since it serves as developer dashboard. If you want to provide a custom admin for your application, then add it as another SPA alongside it, eg. place it inside `/pb_public/admin/` (or embed it in the executable) and it will be available at `/admin/` (btw. that is one of the main reasons why the PocketBase admin route is with underscore so that it doesn't interfere with your app ones).

**Reply by nicetomytyuk** [reply]
@ganigeorgiev Hi there, I notice that echo.HandlerFunc and echo.Context do not exist anymore, what should we use instead?

**Reply by ganigeorgiev** [reply]
@nicetomytyuk PocketBase v0.23+ no longer uses `echo` and instead for routing has a thin wrapper around the Go standard `net/http.ServeMux`.

For more information and examples, please read https://pocketbase.io/docs/go-routing/.

If you are asking for the PocketBase v0.23+ equivalent example of the above code, then that could look something like (_I haven' tested it_):

```go
package main

import (
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
)

func adminRedirect(e *core.RequestEvent) error {
	if strings.HasPrefix(e.Request.URL.Path, "/_/") {
		// admin page -> redirect to your custom url
		return e.Redirect(http.StatusTemporaryRedirect, "https://pocketbase.io")
	}

	return e.Next()
}

func main() {
	app := pocketbase.New()

	app.OnServe().BindFunc(func(se *core.ServeEvent) error {
		se.Router.BindFunc(adminRedirect)

		// serves static files from the provided public dir (if exists)
		se.Router.GET("/{path...}", apis.Static(os.DirFS("/path/to/pb_public"), false))

		return se.Next()
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

But again, as mentioned in my previous comments, this shouldn't really be necessary.

If you have other questions, please don't hijack old discussions and instead consider opening a new one with more details about your use case and what you are trying to do.

**Comment by parkuman** [comment]
I have started a project following ganigeorgiev's advice called PocketCMS! Take a look at the repo here https://github.com/parkuman/pocketcms. Its incredibly simple right now, but it offers a stripped-back admin dashboard that your customers can use to manage content.

**Reply by nicosql** [reply]
> I have started a project following ganigeorgiev's advice called PocketCMS! Take a look at the repo here https://github.com/parkuman/pocketcms. Its incredibly simple right now, but it offers a stripped-back admin dashboard that your customers can use to manage content.

lol - dude, sick - thanks!  gonna give it a go for my company lms

**Comment by jonshipman** [comment]
In general, you should think of the Pocketbase admin as a db administration tool. Similar to phpmyadmin for mysqsl.
